olp-cpp-sdk-core/tests/thread/ContinuationTest.cpp

```diff
@@ -53,145 +56,167 @@ class ContinuationTest : public testing::Test {
 
 TEST_F(ContinuationTest, Failed) {
-  std::promise<olp::client::ApiError> api_error_promise;
+  std::promise<ResponseType<int>> api_error_promise;
   auto future = api_error_promise.get_future();
 
-  auto cont = Create([](olp::thread::ExecutionContext context,
-                        std::function<void(int)>) {
-                context.SetError(olp::client::ApiError::NetworkConnection());
-              }).Finally([&api_error_promise](Response<int> response) {
-    api_error_promise.set_value(response.GetError());
-  });
+  auto continuation =
+      Create([](olp::thread::ExecutionContext context,
+                std::function<void(int)>) {
+        context.SetError(olp::client::ApiError::NetworkConnection());
+      }).Finally([&api_error_promise](ResponseType<int> response) {
+        api_error_promise.set_value(std::move(response));
+      });
+
+  continuation.Run();
 
-  cont.Run();
+  ASSERT_EQ(future.wait_for(kMaxWaitMs), std::future_status::ready);
   auto result = future.get();
 
-  EXPECT_EQ(result.GetErrorCode(), olp::client::ErrorCode::NetworkConnection);
+  ASSERT_EQ(result.GetError().GetErrorCode(),
+            olp::client::ErrorCode::NetworkConnection);
 }
 
 TEST_F(ContinuationTest, NoCrashAfterCallingMethodsAfterRun) {
-  std::promise<olp::client::ApiError> api_error_promise;
+  std::promise<ResponseType<int>> api_error_promise;
   auto future = api_error_promise.get_future();
 
-  auto cont = Create([](olp::thread::ExecutionContext context,
-                        std::function<void(int)>) {
-                context.SetError(olp::client::ApiError::NetworkConnection());
-              }).Finally([&api_error_promise](Response<int> response) {
-    api_error_promise.set_value(response.GetError());
-  });
+  auto continuation =
+      Create([](olp::thread::ExecutionContext context,
+                std::function<void(int)>) {
+        context.SetError(olp::client::ApiError::NetworkConnection());
+      }).Finally([&api_error_promise](ResponseType<int> response) {
+        api_error_promise.set_value(std::move(response));
+      });
 
-  cont.Run();
+  continuation.Run();
+
+  ASSERT_EQ(future.wait_for(kMaxWaitMs), std::future_status::ready);
   auto result = future.get();
 
-  EXPECT_EQ(result.GetErrorCode(), olp::client::ErrorCode::NetworkConnection);
+  ASSERT_EQ(result.GetError().GetErrorCode(),
+            olp::client::ErrorCode::NetworkConnection);
 
-  cont.Then([](olp::thread::ExecutionContext, int, std::function<void(int)>) {
-        FAIL();
+  continuation
+      .Then([](olp::thread::ExecutionContext, int, std::function<void(int)>) {
+        FAIL() << "`Then` method should not be called after the continuation "
+                  "is already run";
       })
-      .Finally([](Response<int>) { FAIL(); });
-  cont.Run();
+      .Finally([](ResponseType<int>) {
+        FAIL() << "`Finally` method should not be called after the "
+                  "continuation is already run";
+      });
+  continuation.Run();
```

Comment:
Run is an async method. Don't you need to ensure somehow that the execution is completely finished?
If there would be some issues in the source code, `Run()` may trigger some tasks, and the tests will crash with SegmentationFault in this case. Such issues are very hard to investigate because SegmentationFault happens when some outer test is already started - it missleads the test that really caused the issue.

Reply:
In this case it checks that Then and Finally aren't called, and Run doesn't cause an exception.